contracts/contracts/interfaces/staking-manager.md

```diff
@@ -0,0 +1,126 @@
+
+## Register Validator
+
+Related ACP-77 section [here](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets#step-1-retrieve-a-bls-multisig-from-the-subnet)
+
+### registerValidator
+
+Register validator is the process of adding a validator to the subnet.The contract initiates this with a `registerValidator` function that takes following inputs:
+
+- Subnet ID (bytes32): The ID of the subnet where the validator is being registered
+- Node ID (bytes32): The ID of the node that is being registered as a validator
+- Amount (uint64): The staked amount that will determine the weight of the validator
+- expirationTime (uint64): The time when this register message will be expired
+- ed25519 signature (bytes): The signature of the message signed by the node's BLS key
+
+The messageID of the request can be crafted with `sha256(subnetID, nodeID, amount, expirationTime, signature)`. This should be same with the one that will be crafted for P-chain.
+
+The contract should verify these:
+
+- (Optional) the staking amount is greater than the minimum staking amount
+- (Optional) The node is not already registered as a validator. This will be handled by the P-Chain though.
+- (Optional) Expiry timestamp is not in the past
+- The signature has valid length (64 bytes)
+- The nodeID is not empty
+- The messageID is not already used
+
+The contract should then send a warp message to P-Chain and lock the staked amount in the contract. The manager should not start accruing reward at this time since the validator is not yet accepted by P-Chain and it's not guaranteed. The warp message will be aggregated and signed, then the validator will issue a `RegisterSubnetValidatorTx` transaction on P-Chain with warp message included. More details from ACP-77 [here](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets#step-2-issue-a-registersubnetvalidatortx-on-the-p-chain)
+
+### receiveRegisterValidator
+
+Once `RegisterSubnetValidatorTx` is accepted a warp message will be aggregated and relayed to this contract as a result of `RegisterSubnetValidatorTx`. Upon receiving and verifying this message (via `receiveRegisterValidator`), the contract should activate the validator and start accruing rewards. The message will only contain `messageID` which corresponds to the `messageID` in `registerValidator` request. The contract should verify these:
+
+- Verified warp message
+- The typecheck on the warp message (== ValidatorRegisteredMessage)
+- messageID is not empty
+- The message is not already used
+- (?) There is a pending request for this messageID
+- (?) The nodeID in the pending request is not a validator already
+- (?) Message is from P-Chain
+
+TODO: we might not want last 3 checks (with ?) to minimize reverts. The contract should be able to handle these cases without reverting. This is because the timing of the messages is not guaranteed and the contract should be able to handle out-of-order messages.
```

Comment:
@michaelkaplan13 

- If there isn't a pending request for this message ID, doesn't that mean we've been sent something invalid?
- The nodeID in the pending request being a validator already seems fine, we may send multiple requests to register.
- How would we handle a message not from the P-Chain here? Shouldn't that revert, as a malicious subnet could send a similar message?

Reply:
1- I guess yes, sending a msg from another subnet should be mostly covered before msg arrives here. But still we'd assume that it's not possible + we assume those pending requests get never deleted. Otherwise one can replay those messages. 
2- in theory we can send many requests, but P-chain won't accept them if there is already a validator with that nodeID. We should definetly let multiple requests to be sent in case there is no validator (to prevent blocking that nodeID), but I was not %100 sure if we should let those if there is a known validator for that nodeID. Mostly to prevent replay attacks. 
3- as told in 1 I was assuming the message would not possibly arrive at this point/not even signed. 

Looking at this again, it now seems that we should probably add all possible checks.